### ISSUE TITLE
Get the best available time counter for node call duration calculation

### ISSUE THREAD

**Issue by smarie** [pytest-dev/pytest#4391]
Hi all

In [pytest-harvest](https://smarie.github.io/python-pytest-harvest/) I provide a function to easily retrieve all information available about test nodes. Among these information, a "duration" is provided (the original from pytest is in seconds, I convert it to ms). However it seems that this number is calculated (in [CallInfo](https://github.com/pytest-dev/pytest/blob/master/src/_pytest/runner.py#L210)) using `time()`, which is not the most precise tool we have to measure durations in python.

Other built-in possibilities are suggested [here](https://stackoverflow.com/questions/3620943/measuring-elapsed-time-with-the-time-module/47637891#47637891). 

A good source of inspiration can also be found in [pytest-benchmark](https://github.com/ionelmc/pytest-benchmark/blob/master/src/pytest_benchmark/timers.py).

Let me know if it makes sense

**Comment by RonnyPfannschmidt** [comment]
sounds like we could switch to `perf_counter` mid 2019 after we branch of the 2.7 bugfixing branch

**Comment by smarie** [comment]
That's a fast and positive answer, awesome !

**Comment by Zac-HD** [comment]
We actually can't *change* without breaking existing code, because the reference point for `perf_counter` is undefined but the `.start` and `.stop` attributes are public API.

However, I would strongly support adding an `.duration` attribute of type `datetime.timedelta`, calculated from `perf_counter` on Python 3 (and falling back to `time` on Python 2).
```python
try:
    from time import perf_counter
except ImportError:
    from time import time as perf_counter

...

        # We store start and end timestamps separately from the duration,
        # to get better resolution and avoid problems from clock resets.
        self.start = time()
        _start = perf_counter()
        try:
            self.result = func()
        except:  # unchanged
        self.duration = timedelta(seconds=perf_counter() - _start)
        self.stop = time()
```

**Comment by smarie** [comment]
You are right. This duration attribute would then need to be used in the report object in place of [this stop - start](https://github.com/pytest-dev/pytest/blob/master/src/_pytest/runner.py#L233)

**Comment by RonnyPfannschmidt** [comment]
as far as im concerned - any monotonic higher resolution time source will do us better than `time.time()`

start/end are both floats and we never defined a point of reference, so its perfectly reasonable to me, to just switch out the data source there without too much of a hassle

**Comment by smarie** [comment]
Hi there, any other feedback on this topic ? I can try a PR one of these weeks if needed.

**Comment by nicoddemus** [comment]
> Also we should maybe check that in python 2.7 there are no better alternatives. For example datetime.now() 

Not sure about `datetime.now()`, but on Windows for example `time.clock()` is more precise... but I agree we can probably wait until we drop 2.7 (which should be soon anyway).

**Comment by RonnyPfannschmidt** [comment]
Anything not a float is a breaking api change

**Comment by nicoddemus** [comment]
> Anything not a float is a breaking api change

Agreed, I was mentioning `datetime.now()` just regarding its precision

**Comment by Zac-HD** [comment]
Better idea: on Python 3, we could track the *duration* using `perf_counter()`, get the `stop` time with `time()`, and then adjust the `start` attribute to match.  Backwards compatible, and resolution limited not by the clock but by float64 (to slightly sub-microsecond precision)!

I'd be happy to accept a PR for this now, or it could wait for pytest 5.0 at which point it wouldn't need a fallback path for Python 2.

**Comment by nicoddemus** [comment]
> Better idea: on Python 3, we could track the duration using perf_counter(), get the stop time with time(), and then adjust the start attribute to match. Backwards compatible, and resolution limited not by the clock but by float64 (to slightly sub-microsecond precision)!

I think we can just go ahead and use `perf_counter` to obtain `duration`, without resorting to `time.time()` at all; we never made any promises about the precision of the `duration` attribute, only about the data type being `float`.

**Comment by nicoddemus** [comment]
~~We don't have `start`/`end` attributes, only a `duration` attribute: https://github.com/pytest-dev/pytest/blob/master/src/_pytest/reports.py#L332~~

Now I understand, thanks, I got confused because I thought we were talking about the duration of the `TestReport` object, not `CallInfo`, my bad. 😅

**Comment by smarie** [comment]
I finally took some time to fix this, see #6939.
